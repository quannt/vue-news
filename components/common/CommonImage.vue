<template>
  <div class="vue-load-image">
    <slot v-if="status === 'loaded'" name="image"/>
    <slot v-else-if="status === 'failed'" name="error"/>
    <slot v-else-if="status === 'loading'" name="preloader"/>
  </div>
</template>

<script>
const Status = {
  PENDING: 'pending',
  LOADING: 'loading',
  LOADED: 'loaded',
  FAILED: 'failed'
}
export default {
  // https://github.com/john015/vue-load-image
  // Could not make this work with nuxt.config.js so I'm using the code from the plugin directly here

  data() {
    return {
      status: null,
      img: null,
      src: null
    }
  },
  watch: {
    src(value) {
      this.status = value ? Status.LOADING : Status.PENDING
    }
  },
  created() {
    this.src =
      this.$slots.image[0].data.attrs.src ||
      this.$slots.image[0].data.attrs['data-src']
    if (this.src) {
      this.status = Status.LOADING
      this.createLoader()
    } else {
      this.status = Status.PENDING
    }
  },
  updated() {
    const receivedSrc =
      this.$slots.image[0].data.attrs.src || this.$slots.image[0].data.attrs
    if (this.status === Status.LOADING && !this.img) {
      this.createLoader()
    } else if (this.src !== receivedSrc) {
      this.src = receivedSrc
      this.createLoader()
    }
  },
  methods: {
    createLoader() {
      this.destroyLoader()
      this.img = new Image()
      this.img.onload = this.handleLoad
      this.img.onerror = this.handleError
      this.img.src = this.src
    },
    destroyLoader() {
      if (this.img) {
        this.img.onload = null
        this.img.onerror = null
        this.img = null
      }
    },
    handleLoad() {
      this.destroyLoader()
      this.status = Status.LOADED
      this.$emit('onLoad')
    },
    handleError(error) {
      this.destroyLoader()
      this.status = Status.FAILED
      this.$emit('onError', error)
    }
  }
}
</script>