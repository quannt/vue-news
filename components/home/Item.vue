<template>
  <div class="article-container">
    <no-ssr>
      <div class="article-image">
        <a :href="article.url">
          <common-image>
            <img slot="image" :src="article.urlToImage" alt class="article-image__image">
            <img slot="preloader" src="~/assets/default-image.png" alt class="article-image__image">
            <img slot="error" src="~/assets/default-image.png" alt class="article-image__image">
          </common-image>
        </a>
      </div>
    </no-ssr>

    <div class="article-body">
      <a :href="article.url">
        <div class="article-body__header">{{ article.title}}</div>
        <div class="article-body__footer">{{ article.source.name }} - {{ timeAgo }}</div>
      </a>
    </div>
  </div>
</template>

<script>
import { format } from 'timeago.js'
import CommonImage from '~/components/common/CommonImage'

export default {
  name: 'Item',
  components: {
    CommonImage
  },
  props: {
    article: {
      type: Object,
      // Object or array defaults must be returned from
      // a factory function
      default: function() {
        return {
          source: {
            id: null,
            name: null
          },
          author: null,
          title: null,
          description: null,
          url: null,
          urlToImage: null,
          publishedAt: null,
          content: null
        }
      }
    }
  },
  data() {
    return {}
  },
  computed: {
    timeAgo() {
      // return format(Date.now() - 11 * 1000 * 60 * 60)
      return format(this.article.publishedAt)
    }
  }
}
</script>

<style>
a {
  text-decoration: none;
}
.article-container {
  display: flex;
  padding: 18px;
}

.article-image {
  width: 120px;
  height: 90px;
  flex-shrink: 0;
}

.article-image__image {
  object-fit: cover;
}

.article-body {
  flex: 1 1 auto;
  padding-left: 18px;
}

.article-body__header {
  font-size: 16px;
  color: #171717;
}

.article-body__footer {
  display: flex;
  font-size: 12px;
  padding-top: 18px;
  color: #707070;
}
</style>
