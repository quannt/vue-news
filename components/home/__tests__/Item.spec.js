import Item from '~/components/home/Item'
import CommonImage from '~/components/common/CommonImage'

import { shallowMount } from '@vue/test-utils'

const article = {
  source: { id: null, name: 'Gizmochina.com' },
  author: null,
  title:
    'Sony Xperia XZ4 AnTuTu score puts it ahead of every phone - gizmochina',
  description:
    'The Xperia XZ4 is Sony‘s upcoming flagship phone for this year and it is expected to pack a punch with the Snapdragon 855 processor under its hood. If you want...',
  url:
    'https://www.gizmochina.com/2019/01/02/sony-xperia-xz4-antutu-score-puts-it-ahead-of-every-phone/',
  urlToImage:
    'https://www.gizmochina.com/wp-content/uploads/2018/11/Sony-Xperia-XZ4-renders-2.jpg',
  publishedAt: '2019-01-02T10:33:00Z',
  content:
    'The Xperia XZ4 is Sony ‘s upcoming flagship phone for this year and it is expected to pack a punch with the Snapdragon 855 processor under its hood. If you want to know just how powerful the processor is inside the phone, then you should check out its AnTuTu … [+1105 chars]'
}

describe('Item.vue', () => {
  test('renders article title', () => {
    const wrapper = shallowMount(Item, {
      propsData: { article }
    })

    expect(wrapper.text()).toContain(article.title)
  })

  test('renders article source name', () => {
    const wrapper = shallowMount(Item, {
      propsData: { article }
    })

    expect(wrapper.text()).toContain(article.source.name)
  })

  test('renders article image', () => {
    const wrapper = shallowMount(Item, {
      propsData: { article }
    })

    const commonImage = wrapper.find(CommonImage)
    expect(commonImage.is(CommonImage)).toBe(true)
  })
})
