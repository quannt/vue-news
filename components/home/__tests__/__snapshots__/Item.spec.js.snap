// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue renders snapshot correctly 1`] = `
<div
  class="article-container"
>
  <no-ssr>
    <div
      class="article-image"
    >
      <a
        href="https://www.gizmochina.com/2019/01/02/sony-xperia-xz4-antutu-score-puts-it-ahead-of-every-phone/"
      >
        <common-image-stub>
          <img
            alt=""
            class="article-image__image"
            src="https://www.gizmochina.com/wp-content/uploads/2018/11/Sony-Xperia-XZ4-renders-2.jpg"
          />
           
          <img
            alt=""
            class="article-image__image"
            src="~/assets/default-image.png"
          />
           
          <img
            alt=""
            class="article-image__image"
            src="~/assets/default-image.png"
          />
        </common-image-stub>
      </a>
    </div>
  </no-ssr>
   
  <div
    class="article-body"
  >
    <a
      href="https://www.gizmochina.com/2019/01/02/sony-xperia-xz4-antutu-score-puts-it-ahead-of-every-phone/"
    >
      <div
        class="article-body__header"
      >
        Sony Xperia XZ4 AnTuTu score puts it ahead of every phone - gizmochina
      </div>
       
      <div
        class="article-body__footer"
      >
        Gizmochina.com - 1 week ago
      </div>
    </a>
  </div>
</div>
`;
